### PR TITLE
Imgseg encoder

### PR DESCRIPTION
**Pull request by jonah-chen**
Implemented the model in the 2018 paper for DeepLab(v3) (NOT the 2017 paper), separated the encoder and decoder.

I plan to use the encoder with a different decoder for this project. Although it may be helpful to pretrain the model.

### CODE REVIEW

**Comment by jonah-chen** [comment]
Added a testing file using the following features:
- Uses the [checkpoint](https://pytorch.org/docs/stable/checkpoint.html#torch.utils.checkpoint.checkpoint) feature in PyTorch to reduce memory usage. Now, model trains on batch size of 16 images (size 512x512) with ~9GB VRAM. This should resolve #12.
- Use multiprocessing (`num_workers=24` in particular) to accelerate the dataloader, increasing GPU usage to over 90% (>400W)
- Put the encoder and decoder into one module with a forward method.
- Uses `tqdm` to show progress bars. One epoch (186 mini-batch) takes about 90 seconds.

There are still a few things that are not perfect.
- CPU usage is a little high, likely caused by the data augmentation that is going on for this test. I don't think this is a major issue, as the preprocessed data do not require too much more augmentation. Note that increasing `num_workers` from 0 to 24 increased GPU usage from 10% to over 90%, which is good. 
- I have to use `.cuda()` function otherwise the code will error out. I think this is because the `torch.set_default_tensor_type` is not as universal which leaves some tensors in the host (CPU) memory.


#### There are a few logs that are slightly concerning, hence I will post them for the record.
```
.../torch/utils/checkpoint.py:25: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
```
Not sure what's going on with this one, but the loss does go down with training on one batch or training on the full dataset. 

```
.../torch/nn/functional.py:3454: UserWarning: Default upsampling behavior when mode=bilinear is changed to align_corners=False since 0.4.0. Please specify align_corners=True if the old behavior is desired. See the documentation of nn.Upsample for details.
  warnings.warn(
```
No clue what this means.

**Comment by jonah-chen** [comment]
We don't need this anymore
